### PR TITLE
DROOLS-7436: Allow to use groupBy in drl

### PR DESCRIPTION
**Pull request by Christopher-Chianelli**
Extends the DRL grammar to allow access to groupby:

```yacc
    lhsGroupBy ::= 'groupby' '(' 
                                 /* Source pattern, what is being grouped */  
                                 lhsAnd ';'
                                 /* grouping function, applied to each item the source pattern matches */
                                 /* there will always be exactly one grouping function */
                                 [identifier ':'] expression ';'
                                 /* accumulators (at least one) */
                                 accumulateFunctionBinding (COMMA accumulateFunctionBinding)*
                                 /* include the result only if all the constraints matches (optional) */
                                 [';' CONSTRAINTS]
                             ')'
```

Currently a work in progress; some tests in `GroupByTest` fail when converted to DRL due to a few separate issues: 

- the key variable being hidden when an outer group by uses an inner group by
- duplicate parameter/variable when only a single variable used

Additionally, some tests in  `GroupByTest` cannot be faithfully rewritten using DRL, because they use groupby without any accumulators.

**JIRA**:

[link](https://issues.redhat.com/browse/DROOLS-7436)

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request checks</b>  
  Please add comment: <b>Jenkins retest this</b>

- for a <b>specific pull request check</b>  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] tests</b>

- for a <b>full downstream build</b> 
  - for <b>jenkins</b> job: please add comment: <b>Jenkins run fdb</b>
  - for <b>github actions</b> job: add the label `run_fdb`

- <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

- <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

- <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>

- for <b>quarkus branch checks</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins run quarkus-branch</b>

- for a <b>quarkus branch specific check</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] quarkus-branch</b>

- for <b>quarkus main checks</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins run quarkus-main</b>

- for a <b>specific quarkus main check</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] quarkus-main</b>

- for <b>quarkus lts checks</b>  
  Run checks against Quarkus lts branch  
  Please add comment: <b>Jenkins run quarkus-lts</b>

- for a <b>specific quarkus lts check</b>  
  Run checks against Quarkus lts branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] quarkus-lts</b>

- for <b>native checks</b>  
  Run native checks  
  Please add comment: <b>Jenkins run native</b>

- for a <b>specific native check</b>  
  Run native checks 
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] native</b>

 - for <b>native lts checks</b>  
  Run native checks against quarkus lts branch
  Please add comment: <b>Jenkins run native-lts</b>

- for a <b>specific native lts check</b>  
  Run native checks against quarkus lts branch
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] native-lts</b>
</details>

<details>
<summary>
How to backport a pull request to a different branch?
</summary>

In order to automatically create a **backporting pull request** please add one or more labels having the following format `backport-<branch-name>`, where `<branch-name>` is the name of the branch where the pull request must be backported to (e.g., `backport-7.67.x` to backport the original PR to the `7.67.x` branch).

> **NOTE**: **backporting** is an action aiming to move a change (usually a commit) from a branch (usually the main one) to another one, which is generally referring to a still maintained release branch. Keeping it simple: it is about to move a specific change or a set of them from one branch to another.

Once the original pull request is successfully merged, the automated action will create one backporting pull request per each label (with the previous format) that has been added.

If something goes wrong, the author will be notified and at this point a manual backporting is needed.

> **NOTE**: this automated backporting is triggered whenever a pull request on `main` branch is labeled or closed, but both conditions must be satisfied to get the new PR created.
</details>

### CODE REVIEW

**Comment by mariofusco** [comment]
> Additionally, some tests in GroupByTest cannot be faithfully rewritten using DRL, because they use groupby without any accumulators.

The use case of a groupby with no accumulators was expressly added for Optaplanner to cover the situation when you're only interested in the groups' keys but not in any cumulative value of the grouped facts. Note however that in plain drl the same result can be more easily obtained with a [collectSet](https://github.com/kiegroup/drools/blob/main/drools-core/src/main/java/org/drools/core/base/accumulators/CollectSetAccumulateFunction.java). 

For uniformity you can also try to make the accumulate function(s) totally optionally in the groupby clause, but if this creates any problem or ambiguity in the drl parser please feel free to ignore this issue and rewrite the tests using the groupby with no accumulators with a simple collectSet.

**Comment by Christopher-Chianelli** [comment]
> For uniformity you can also try to make the accumulate function(s) totally optionally in the groupby clause, but if this creates any problem or ambiguity in the drl parser please feel free to ignore this issue and rewrite the tests using the groupby with no accumulators with a simple collectSet.

Making the accumulate optional would introduce ambiguity in the DRL parser; in particular the code

```
groupby($p: Person($name: name); $name; function($p))
```

would match

```
'groupby' '(' source ';' grouper ';' constraints ')'
```

and


```
'groupby' '(' source ';' grouper ';' accumulator ')'
```
.

This ambiguity could be solved with a variety of grammar modification:

Option A: Require two semicolons to specify constraints with no accumulator present
- This IMO make the code for constraints without accumulator look ugly, but this will probably be an uncommon operation and doesn't present any other issue.

Option B: Use a comma as a separator between the grouping function and accumulators
-  This works because there will be exactly one grouping function. However, this make the distinction between the grouping function and accumulators less clear, and prevents modifying the grammar in the future to allow multiple grouping function in a backward compatible way (without introducing two separate ways to create groupby).

Option C: Status Quo
- The existence of [collectSet](https://github.com/kiegroup/drools/blob/main/drools-core/src/main/java/org/drools/core/base/accumulators/CollectSetAccumulateFunction.java) means groupby without accumulators is just an alternative for users. However, this does not solve the testing problem; the tests are for groupby with no accumulators, NOT an accumulate using CollectSet (although these two should have the same result). This is important because OptaPlanner uses groupby with no accumulator, and changing the test to use accumulate with CollectSet risks a regression not being caught.

**Comment by Christopher-Chianelli** [comment]
Note: the tests should be reviewed for accuracy; For instance, I used `$var : Type() from eval(expression)` in order to do some bindings when translating tests that use binding on the groupKey (and thus cannot be done with something like `Person($name: name)`.

File: drools-model/drools-model-codegen/src/main/java/org/drools/model/codegen/execmodel/generator/visitor/accumulate/GroupByVisitor.java [hunk 1]
```diff
@@ -64,11 +71,18 @@ protected void processAccumulateFunctions(AccumulateDescr descr, PatternDescr ba
...
         accumulateDSL.addArgument(toVar(groupingKey));
 
         accumulateDSL.addArgument( buildConstraintExpression(typedExpression.getExpression(), result.getUsedDeclarations()) );
 
         super.processAccumulateFunctions(descr, basePattern, input, accumulateDSL);
     }
+
+    private void addGroupKeyBindingAsDeclaration(RuleContext context, String bindingId, TypedExpression grouppingFunction) {
+        if (bindingId != null) {
+            Class grouppingFunctionResultType = grouppingFunction.getRawClass();
```
Comment (mariofusco) [inline drools-model/drools-model-codegen/src/main/java/org/drools/model/codegen/execmodel/generator/visitor/accumulate/GroupByVisitor.java]: Typo: grou**pp**ing with 2 Ps

**Review by mariofusco — APPROVED** [review]

**Pull request merged**